### PR TITLE
Main API change for v0.11 🎂 

### LINKED ISSUE #91: [RFC] `v0.11` breaking change to main API
There are mainly three ways a user may want to construct a histogram:

1. Directly construct a histogram from known `binedges, bincounts, sumw2s, nentries` (e.g. loading histogram from ROOT)
2. Starting with a known histogram bin count type and `binedges`, then `push!()` into this histogram\
3. Starting with a bunch of data already in array (possibly with `weights` as well, maybe already know desired `binedges`)

Our current APIs are quite confusing, I want to re-design it, but there are ~maybe binary-choice~ some design questions we have to answer, I will say what they are and write down what I prefer for posterity:

## A) `fit(Hist1D, ...)` vs. B) `Hist1D(...)`

The first obvious question is do we want to use `Hist1D(...)` for everything or reserve it for constructor-like use case only. I prefer option B) because even if we use `fit()` for the 3nd kind of use case, use case 1. and 2. will still share `Hist1D(...)` api,

## Positional args vs. Keyword args
I would call use cases 1. and 2. "constructor-like" where 3. is "fit-like". Here we have one obvious choice (or the opposite), I have been thinking about something like this:

```julia
# constructor-like
Hist1D(; eltye::Type{T}, binedges, bincounts, sumw2s, nentries, overflow) where T

# fit-like
Hist1D(ary, weights, binedges; eltype::Type{T}, overflow) where T
```

we can as easily exchange these two, make `fit-like` use keyword argument. But somehow I feel like make constructir-like use case 100% explicit might be more natural

### PR DESCRIPTION
**Pull request by Moelf**
close #91

- [x] implement the proposed APIs
- [x] fix tests accordingly
- [x] add breaking change note at the top of README
- [x] add new doc page for the overview https://moelf.github.io/FHist.jl/previews/PR95/start/
- [x] add a doc page describing the 1D vs. multi dimension cases
- [x] read all existing docs see if there's inconsistency
- [x] fix plotting extension and notebook
- [x] re-implement #87 somehow

### CODE REVIEW

File: src/FHist.jl [hunk 1]
```diff
@@ -16,34 +16,135 @@ using Base.Threads: SpinLock
...
 
 using BayesHistogram
 export BayesHistogram
 
 const _default_overflow = false
+_to_tuple(x::Tuple) = x
+_to_tuple(x) = tuple(x)
 
+_from_tuple(x::Tuple{Any}) = only(x)
+_from_tuple(x) = x
```
Comment (Moelf) [inline src/FHist.jl:28]: we special case `Hist1D`, in the sense that the underlying data structure for `binedges` is still `NTuple{1, AbstractVector}`, but for all the user-facing APIs:
- `binedges`
- `bincenters`

we return just an `AbstractVector`, similarly, for constructor, we take just an `AbstractVector` as well

**Comment by Moelf** [comment]
@tamasgal @jjgomezcadenas @bojohnson5 @peremato

Can you read this single page of the doc where I try to explain the big picture, and let me know if it makes sense / clean?

https://moelf.github.io/FHist.jl/previews/PR95/start/

File: docs/src/start.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+```
+
+There are two possible ways to make a histogram:
+- make an empty histogram to be filled later (via [`atomic_push!`](@ref))
+- make a filled histogram given data
+
+We call the first one "constructor like" use case and the second one "fit like" use case (as in "fit
+data to a histogram").
+
+Each of the two options imply a set of options that can go with them. The the two sets of options
```
Comment (bojohnson5) [inline docs/src/start.md:19]: Then the two sets

**Comment by bojohnson5** [comment]
I really like the overview page, noticed a single typo on first read-through that I added a comment on. It makes sense and gives a good idea of the two methods for making a histogram.

**Comment by Moelf** [comment]
I also added: https://moelf.github.io/FHist.jl/previews/PR95/usage/ which is not complete and not very beautiful, but it's a strict improvement over what we had before.

I also proof read it, seems fine to me

**Comment by Moelf** [comment]
ok I rebased to the 5 commits I want to keep for future reference (partial revert if we have major issues need do over)

**Comment by tamasgal** [comment]
Very nice! I need your opinion on the tiny mess I want to fix with https://github.com/Moelf/FHist.jl/pull/96 which you could rebase on as well...

**Comment by jjgomezcadenas** [comment]
Hi Jerry, the doc page is very nice. For reference. In the good old times (please bear with me)  atomic push was the only way to fill an histogram (hbook then hfill). In fact, back then, the nice thing about histos was that they were expandable vectors, when we all programmed in Fortran which had fixed-size arrays. So, actually, the logic of filling the histogram was, in practice, "fill a vector of unknown size, then bin it". 

Which brings me to the need (or lack of) atomic push in a language like julia, where it is trivial to fill a vector, then histogram it. Of course atomic push doesn't hurt, and one can argue that booking histograms before hand may allow the use to keep them as some kind of high level data. In any case, good work!

**Review by tamasgal — APPROVED** [review]

**Comment by peremato** [comment]
Hi Jerry,
Nice improvements. I have a couple of comments:
- I find always complicated to provide binedges instead of the 'traditional' ROOT (min, max, nbins). What I most of the time know is the number of bins I want to have. Then, your your API I have to figure out what I need to put in the increment to obtain them.
- We discussed  the need to automatic setting of the min, max and the number of bins, even for the case that you populate the histogram one-by-one.
- The other issue was to have a predictable concrete type for the histogram. This is the case when you want to put histograms in user data structures. If the type is not concrete, the overhead generated by the the dynamic dispatch is larger probably than the push! function itself.

**Comment by Moelf** [comment]
On the topic of auto resizing histogram, I want to understand what does ROOT have and what is actually the most useful in practice, for example, I know of the design in https://github.com/joshday/OnlineStats.jl/blob/b365492a630e4d00f64707194e715df891f84888/src/stats/histograms.jl#L130; they basically do `rebin(hist, 2)` so that the `maximum(binedges(hist))` doubles every time, until they can accommodate new data. Is this kind of approach what you want? I think in theory we don't need a new type for this, just need a new `expand_push!()` function to signal you're willing to mutate histogram `binedges` if the data doesn't fit

**Comment by peremato** [comment]
Hi Jerry. Using your branch I ses the following:
- It is true that I can use `binedges=range(min, max,nbis+1)` but is not friendly I think. What about defining an new constructor like this:
  ```
  Hist1D(::Type{T},min,max,nbins) where T = Hist1D(;counttype=T,binedges=range(min,max,nbins+1))
  ```
  Then I can just do `Hist1D(Float64,-2,2,10)`
- The actual type is still a lengthy type.
  ```
  julia> typeof(Hist1D(Int32,-2,2,10))
  Hist1D{Int32, Tuple{StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}}
  ```
- Yes, filling an histogram without the bins defined, your need probably to buffer temporary a few entries and then do the binning. I do not know how ROOT does it.

**Comment by Moelf** [comment]
>`Hist1D(::Type{T},min,max,nbins) where T = Hist1D(;counttype=T,binedges=range(min,max,nbins+1))`

That's a no-go, the biggest reason we want to do this re-design is to make the "with data" and "without data" completely clean. In this case, because you're making an empty histogram, it must not use a positional argument. It left with us the possibility of adding `; min, max` to the list of keyword arguments, but we will have to do sanity check and friendly error message when it conflicts with `binedges=...`

Actually, I just noticed **all you need** is `range(min, max, length+1)`, you don't need keyword argument.

- - -
 
Now for the types, it's somewhat inevitable because Julia does high-precision handling for ranges involving floating point:
```julia
julia> typeof(range(-2, 2, 11))
StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}

julia> typeof(1:2:5)
StepRange{Int64, Int64}

julia> typeof(1:3)
UnitRange{Int64}
```

I think the only practical option is to do something analogue to what ROOT does:https://github.com/JuliaHEP/UnROOT.jl/blob/d0628a90517ab34cfc3ffc301f538ae1d803a9bc/src/utils.jl#L110-L118

Basically have fields for both cases and dynamically pick the available one to use (maybe based on a flag) on the fly. I don't like this based on aesthetics and potential confusion for e.g. serialization.

**Comment by peremato** [comment]
Yes perhaps is not very nice to have positional arguments, but I do not understand that with a 'clean' design to mean `positional=with data` and `keyword=empty`. 
- Would it be nice to have `Hist1D(::Type{T},...)` for empty and `Hist1D(a::AbstractVector{T},...)` to distinguish them?  
- Then, instead of using `binedges` what about two use `Axis(min, max, nbins)`. I think this is the design of Boost histograms in Python. 
  ``` 
  ax = bh.axes.Regular(100, start=-5, stop=5) 
  hist = bh.Histogram(ax)
  hist.fill(np.random.randn(1_000_000))
  ```

**Comment by Moelf** [comment]
I actually think it's unlikely for someone to customize `counttype` given `Float64` can represent integers exactly up to `9e15`. But to translate that Python example

```python
ax = bh.axes.Regular(100, start=-5, stop=5) 
hist = bh.Histogram(ax)
hist.fill(np.random.randn(1_000_000))
```

```julia
binedges = range(-5, 5, 101)
hist = Hist1D(; binedges)
append!(hist, randn(1_000_000))
```

**Comment by Moelf** [comment]
I think we can almost merge this now. If you look at our CI coverage, it's pretty impressive, however, I'm certain we have holes in functions in the sense that some features are not implemented uniformely in 1D, 2D, 3D

**Comment by peremato** [comment]
Hi @Moelf. Sorry to insist on resulting concrete type. I believe this is a very common use case. People will have arrays of histograms of all sorts, or embedded in their own data structures. Please have a look at the following benchmark.
```
using FHist
using BenchmarkTools

h = Hist1D(binedges=-4.5:0.5:5.)

const data = randn(100000)

struct MyData
    h::Hist1D
end

struct MyData2
    h::Hist1D{Float64, Tuple{StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}}
end

function hcreate(d)
    Hist1D(d)
end

function hfill(h,d)
    empty!(h)
    for v in d
        push!(h,v)
    end
    h
end

mydata = MyData(Hist1D(binedges=-4.5:0.5:5.))
mydata2 = MyData2(Hist1D(binedges=-4.5:0.5:5.))

function sfill(s, d)
    empty!(s.h)
    for v in d
        push!(s.h, v)
    end
    s.h
end

println("Creating histogram from vector:")
@btime hcreate($data)
println("filling an existing histogram")
@btime hfill($h, $data)
println("filling an abstrcat hitogram in user structure")
@btime sfill($mydata, $data)
println("filling a concrete histogram in user structure")
@btime sfill($mydata2, $data)
```
gives the following results:
```
Creating histogram from vector:
  278.125 μs (7 allocations: 624 bytes)
filling an existing histogram
  145.459 μs (0 allocations: 0 bytes)
filling an abstract histogram in user structure
  1.470 ms (100000 allocations: 1.53 MiB)
filling a concrete histogram in user structure
  145.209 μs (0 allocations: 0 bytes)
```
If you do to specify the type with all full glory of details you have a 10x less performance. The other strange thing is that creating and filling from a vector is 2x slower than filling after creation.

**Comment by Moelf** [comment]
Don't worry about it, I'm 💯 with the use case, I just don't know how to do it. If anyone have suggestions, we can try it. I think I can summarize the problems as the following, before any coding challenge, here are the logical difficulties:

- if we want to remove parametric `counttype`, what should `bincounts(h::Hist1D)` be, `Float64`? Is that going to meet all HEP needs?
- if we want to remove parametric `E` (for `binedges`), what should it be?

The second question is particularly problematic, let's give the `binedges` its own type to help visualize what possibilities we even have.

### Current Design:
```julia
struct BinEdges{E<:AbstractVector}
    binedges::E
end
```

the practical reason is that `[1, 10, 1000]` and `1:100` and `1:5:100` are completely different types, with the last two `<: AbstractRange` thus enables fast bin index look up. Now let's try to get rid of `E`

### Possibility 1: union splitting on reduced set of types
```julia
T = typeof(1:1.0:3)
# StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}

struct BinEdges
    binedges::Union{Vector{Float64}, T}
end
```

In this design, we:

- collapse all non-uniform binning to `Vector{Float64}`, `[1, 10, 100]` is now `[1.0, 10.0, 100.0]`
- collapse all uniform binning to have type `T`, essentially converting `1:10` to `1.0:1.0:100.0`

and we 🙏 union splitting gives us back the performance.

### Possibility 2: value-based polymorphism
```julia
T = typeof(1:1.0:3)

struct BinEdges
    binedges1::Vector{Float64}
    binedges2::T # defaults to something like -1.0:0.0:1.0
end
```
And each `push!()` we check if `isempty(binedges2)` and decide which "bin index look up" algorithm to use.

Other variations of Possibility 2 include storing `start::Float64, stop::Float64, step::Float64` and check if `iszero(step)` for example.

and we 🙏 runtime splitting is fast enough

**Comment by peremato** [comment]
I do not think the count type is a big deal. We are used in ROOT to indicate if the counts are double, single precision or integers. So something like `Hist1D{Float64}` is not too bad and people would understand. 
Concerning the BinEdges you could predefine a number of them for uniform and non-uniform, etc. In this case the user can specify `Hist1D{Float32, UniformBins}` for example. This is much better that `Hist1D{Float64, Tuple{StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}}`

**Comment by Moelf** [comment]
I did not pick up that you were suggesting a type alias.

One reason to not add and export type alias from `FHist.jl` is because it may be confusing to user.

```julia
julia> const UniformBins = typeof(0:0.1:1);

julia> 0:0.1:1 isa UniformBins
true

julia> 0:10 isa UniformBins
false

julia> UniformBins
StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}
```

The confusion is 2.5 folds, I think:
- the printing will be confusing, `UniformBins` suggest a special type, but users will never be able to see them in any "display" output
- the semantics will be confusing, as you can see some clearly uniform bins will not be uniforms bins. This matters because you can't put a histogram `0:10` into your struct or vector if you instantiated the type with the other one.
- users may be tempted to dispatch on it / specialize without noticing it's actually a Base type

**Comment by peremato** [comment]
No, I would hide in a real struct.
```
struct UniformBins
  binedges::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}
end
```
and forward the methods you call from it.

**Comment by Moelf** [comment]
Conceptually, this doesn't do much because you still have 3 different types for 3 different Julia concepts. If we ignore the printing and user specializing on it for a moment because type wrapper solve those two problems, we still have three types:

- `Our_StepRangeLen`
- `Our_UnitRange`
- `Vector{Float64}`

Now we have defined our own type that is effectively `StepRangeLen{....}`, as a user, they still need to predict the future if they will use (the equivalent of) `1:10` or `1:1.0:10` or `Vector{Float64}` as bin edges.

**Comment by Moelf** [comment]
In fact, if we're going to make our own type for bin edges, I would rather go with Possibility 2 in https://github.com/Moelf/FHist.jl/pull/95#issuecomment-1866089822 since then we can completely get rid of the second parametric type

**Comment by peremato** [comment]
I am not sure if we are agreeing or disagreeing. The user needs to predict he wants, uniform vs. non-uniform. I do not think the case `1:10` is very often used. 
```
julia> UniformBins(1:10)
UniformBins(1.0:1.0:10.0)

julia> UniformBins(1.:1.:10.)
UniformBins(1.0:1.0:10.0)

julia> UniformBins([1.,3.,5.])
ERROR: MethodError: Cannot `convert` an object of type Vector{Float64} to an object of type StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}
```
If you have foreseen uniform and provide a `Vector{Float64}` is your problem. The possibility 2 I do not think is necessary. You do not need both implementations at the same time. 

struct Hist1D{T,E}
  binedges::E
  <different constructors that creates the Hist1D with the right parameters>
end

**Comment by Moelf** [comment]
I guess I'm observing the fact that, without reducing the number of parameters to `Hist1D{...}`, it does not seem to fundamentally make life easier. Basically, to me it doesn't seem to make enough difference between:

>The user needs to predict he wants, uniform vs. non-uniform.
```julia
Hist1D{Float64, UniformBins}
# and
Hist1D{Float64, Vector{Float64}}
Hist1D{Float64, Vector{Int}}
...
```

doesn't appear to be so much better than what we have now:
```julia
UniformBins = typeof(1:1.0:3)
Hist1D{Float64, UniformBins}
# and
Hist1D{Float64, UnitRange{Int}}
Hist1D{Float64, Vector{Float64}}
Hist1D{Float64, Vector{Int}}
...
```

Basically, I guess we kinda have collapsed away two options into one, by making the constructor auto `convert()` between UnitRange and `StepRangeLen`.

In your opinion this is much better?

**Comment by peremato** [comment]
Do you really want to show, and force the users to know how you implemented the histograms? For example, with what you propose the user needs to know that the variable size bins are implemented as Vector{Float64}, and so on. Why not to hide this behind a couple or three axis or binedges types? You can then have the freedom to implement as you wish and the possibility to extending with more sophisticated ones.

**Comment by Moelf** [comment]
I think we agree on that but the conclusion is a bit different, I'm thinking if user still need to differentiate between:
```julia
Hist1D{Float64, UniformBins}
# and
Hist1D{Float64, NonUniformBins}
```

it's not much better, essentially they still need to know:
- histograms are parametric types with 2 parameters
- the first parameter is `counttype`
- the second parameter is `binedges`

Then, I think `NonUniformBins` and `Vector{Float64}` is not that much of a difference.

I would say if we really want to improve this, we should just get rid of the second argument, that would be enough improvements to warrant the engineering effort

**Comment by Moelf** [comment]
btw, another problem you may not be thinking about yet is how does this generalize to 2D histograms:

```julia
Hist2D{Float64, {UniformBins, UniformBins}}
Hist1D{Float64, {NonUniformBins, NonUniformBins}}
Hist1D{Float64, {UniformBins, NonUniformBins}}
Hist1D{Float64, {NonUniformBins, UniformBins}}
```

and so on, if we can get rid of the entire 2nd type parameter, it would really be an improvement

**Comment by Moelf** [comment]
Oh and I think having shit load of parameters is historical:
https://github.com/JuliaStats/StatsBase.jl/blob/4c9e55929eb8409de0e3dd1b03d22b7db62e0cda/src/hist.jl#L186

in any case, I'm currently working on polymorphic binedges so that we only need:
```julia
Hist1D{T}
Hist2D{T}
...
```

we will see how that goes

**Comment by Moelf** [comment]
@peremato tests pass now, I will look into performance later, what do you think of this design? all the meat are in:
- https://github.com/Moelf/FHist.jl/pull/95/files#diff-68d8cd9c20f52c4de45cc18cef09bf3e26c42b374efd4d42fa1db4c88440b1eb

**Comment by peremato** [comment]
Looks very nice. Thanks a lot!

File: src/polybinedges.jl [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            end
+            if any(x -> abs(x) > maxintfloat(Float64), edges)
+                throw(ArgumentError("BinEdges cannot contain element with absolute value larger than $(maxintfloat(Float64))"))
+            end
+            new(false, edges, 0.0:-1.0, Inf, -Inf)
+        end
+    end
+end
+
+Base.@constprop :aggressive isuniform(b::BinEdges) = b.isuniform
```
Comment (peremato) [inline src/polybinedges.jl:36]: @Moelf  I am curious. What is the effect of this `@constprop` ? where do you gain performance?

Comment (Moelf) [reply src/polybinedges.jl:36]: If the constant is propagated, then the if-else will be selected at compile time basically (i.e. which searchsorted method)

But yeah I should verify this is actually doing something

**Comment by Moelf** [comment]
## Example 1
```julia
julia> struct H
           flag::Bool
           val1::Int
           val2::Float64
       end

julia> isuniform(H) = H.flag

julia> Base.@constprop :none getval(H) = isuniform(H) ? H.val1 : H.val2

julia> function main()
           h1 = H(false, 1, 2.0)
           return getval(h1)
       end

julia> main()
2.0

julia> @code_llvm main()
;  @ REPL[4]:1 within `main`
define { {}*, i8 } @julia_main_309([8 x i8]* noalias nocapture noundef nonnull align 8 dereferenceable(8) %0) #0 {
post_union_move:
;  @ REPL[4]:3 within `main`
  %1 = bitcast [8 x i8]* %0 to i64*
  store i64 4611686018427387904, i64* %1, align 8
  ret { {}*, i8 } { {}* null, i8 1 }
}
```


## Example 2
```julia

julia> getval(H) = isuniform(H) ? H.val1 : H.val2
getval (generic function with 1 method)

julia> @code_llvm main()
;  @ REPL[4]:1 within `main`
define double @julia_main_318() #0 {
top:
;  @ REPL[4]:3 within `main`
  ret double 2.000000e+00
}
```

**Comment by Moelf** [comment]
looks like if anything, we should add constant propagation to the caller of `isuniform`
